src/codec.rs

```diff
@@ -1089,24 +1078,51 @@ impl Decode for String {
 }
 
 impl<T: Encode> Encode for [T] {
+	fn size_hint(&self) -> usize {
+		if let IsU8::Yes(_) = <T as Encode>::IS_U8 {
+			self.len() + mem::size_of::<u32>()
+		} else {
+			0
+		}
+	}
+
 	fn encode_to<W: Output>(&self, dest: &mut W) {
 		let len = self.len();
 		assert!(len <= u32::max_value() as usize, "Attempted to serialize a collection with too many elements.");
 		Compact(len as u32).encode_to(dest);
-		for item in self {
-			item.encode_to(dest);
+		if let IsU8::Yes(_) = <T as Encode>::IS_U8 {
```

Comment:
We might want a comment here explaining that it is important to guarantee that this check can be optimized away by the compiler at compile-time.

Reply:
I'm not sure it is that important, having a if condition doesn't add that much overhead.

Also it is ensure that this check is optimized away because T cannot be dynamic. But this is quite unfortunate to be honest.